plugins/README.adoc

```diff
@@ -7,3 +7,256 @@ NOTE: Work in Progress...
 * Use https://pep8.org/[PEP8].
 * File headers and functions should have comments for their intent.
+
+
+== All plugin types should use argspec validation, inbound and outbound
+[%collapsible]
+===
+Explanations:: A plugin's `argument_spec` (or "argspec") defines the supported arguments for a module, as well as their type, defaults and more.
+
+Rationale:: Validating the argspec at the time that it enters and exits the code ensures that these inputs and outputs conform to the plugin's documentation and the user's expectations.
+
+Examples::
+From the https://github.com/ansible-collections/ansible.utils/blob/2.6.0/plugins/filter/ipaddr.py[ipaddr] filter plugin.
+[source,python]
+----
+from ansible_collections.ansible.utils.plugins.module_utils.common.argspec_validate import (
+    AnsibleArgSpecValidator,
+)
+
+
+aav = AnsibleArgSpecValidator(
+    data=data, schema=DOCUMENTATION, name="ipaddr"
+)
+valid, errors, updated_data = aav.validate()
+if not valid:
+    raise AnsibleFilterError(errors)
+return ipaddr(**updated_data)
+----
+====
+
+== All plugin types should have documentation
+[%collapsible]
+====
+Explanations::
+All plugins, regardless of type, should have documentation that describes the input parameters, outputs, and practical examples of how to use it.
+====
+
+== Use sphinx (reST) formatted docstrings
+[%collapsible]
+====
+Explanations::
+All classes, private and public functions should have a sphinx (reST) formatted docstring including all parameters, yields, raises, or returns.
+
+Rationale::
+PEP-257 states that all modules should normally have docstrings, and all functions and classes exported by a module should also have docstrings. Public methods (including the __init__ constructor) should also have docstrings. A package may be documented in the module docstring of the __init__.py file in the package directory.
+
+Examples::
+[source,python]
+----
+"""[Summary]
+
+:param [ParamName]: [ParamDescription], defaults to [DefaultParamVal]
+:type [ParamName]: [ParamType](, optional)
+...
+:raises [ErrorType]: [ErrorDescription]
+...
+:return: [ReturnDescription]
+:rtype: [ReturnType]
+"""
+----
+
+====
+
+== If Python 3 or greater, all code should have type hints and be validated with mypy
+[%collapsible]
+====
+Explanations:: Use Python type hints to document variable types.
+
+Rationale::  Type hints communicate what type a variable can be expected to be in the code. They can be consumed by static analysis tools to ensure that variable usage is consistent within the code base.
+MyPy is a static type checker/
+
+Examples::
+[source,python]
+----
+def greeting(name: str) -> str:
+    return 'Hello ' + name
+----
+
+====
+
+== The use of unittest is discouraged, use pytest instead.
+[%collapsible]
+====
+Explanations:: Use pytest for writing unit tests for plugins
+
+Rationale:: Pytest is the testing framework used by Ansible Engineering and will provide the best experience for plugin developers
+
+Examples::
```

Comment:
We don't have anything in the docs that specifically discusses preferences between unit testing frameworks (and there are certainly many in use in the existing code).  The closest documentation I can link here is the general "how to unit test ansible modules" [doc](https://docs.ansible.com/ansible/latest/dev_guide/testing_units_modules.html) and I'll be the first to admit that it's not for the faint of heart.

We want to encourage plugin developers to write tests, but this is a more advanced developer topic.

Do you believe there would be value in keeping this type of recommendation here if we can expand appropriately on it?  Or would it be better to have a general recommendation for unit testing that links to the devel guide, without prescribing a particular framework?

Reply:
These are all good questions to which I might not have a definitive answer:

1. plugin developement is definitely something we want to have in, and it's an advanced topic where I'm a rookie, so my level of knowledge might not be the bar to measure
2. a bit more rationale might be good, it would be super to understand why the developers use it, but just explaining that the documentation is built around this might be sufficient
3. a link to the official documentation would be of course good

Reply:
The dev guide has a whole Why, When, and How for unit tests so I'll note that info can be found there link out.